sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -263,4 +280,107 @@ case class InsertIntoHadoopFsRelationCommand(
       }
     }.toMap
   }
+
+  /**
+   * Detect the conflict when there are several InsertIntoHadoopFsRelation operations
+   * write to same partition in the same table or a non-partitioned table concurrently.
+   */
+  private def detectConflict(
+      fs: FileSystem,
+      path: Path,
+      staticPartitionKVs: Seq[(String, String)]): Unit = {
+    for (i <- 0 until partitionColumns.size) {
+      Some(".spark-staging-" + i)
+        .map(stagingPath => new Path(path, stagingPath))
+        .foreach { stagingDir =>
+          if (fs.exists(stagingDir)) {
```

Comment:
Lots of `.exists` calls can slow things down... hopefully that won't be a problem here (since listing is not a good alternative thanks to S3).

Reply:
In fact, there will not be a lot of partition columns in production env.
And we usually need invoke `.exists` at most num of partition columns times for a query if no conflict was detected.

Reply:
> (since listing is not a good alternative thanks to S3)
moot for s3 as neither FileOutputCommitter should be used there; does matter for the other stores.

you might want to skip that double check when stagingDir doesn't exist .. it's rechecked on line 351